src/snowflake/snowpark/mock/_connection.py

```diff
@@ -145,6 +145,13 @@ def write_table(
                     raise SnowparkSQLException(f"Table {name} already exists")
                 else:
                     self.table_registry[name] = table
+            elif mode == SaveMode.TRUNCATE:
+                if name in self.table_registry:
+                    target_table = self.table_registry[name]
+                    if table.columns != target_table.columns:
+                        raise SnowparkSQLException("Column mismatch detected")
+
```

Comment:
@sfc-gh-stan / @sfc-gh-aling How should the error string be same as what is raised by server?

Reply:
In that case, I'm happy with this error change.